src/librustc/sir.rs

```diff
@@ -1,174 +1,112 @@
-        for func in &self.funcs {
-            writeln!(dest, "{}", func)?;
+impl SirFuncCx {
+    pub fn new<'tcx>(tcx: TyCtxt<'tcx>, instance: &Instance<'tcx>, num_blocks: usize) -> Self {
+        let symbol_name = format!("new__{}", tcx.symbol_name(*instance).name.as_str());
+
+        let mut flags = 0;
+        for attr in tcx.get_attrs(instance.def_id()).iter() {
+            if attr.check_name(sym::trace_head) {
+                flags |= ykpack::bodyflags::TRACE_HEAD;
+            } else if attr.check_name(sym::trace_tail) {
+                flags |= ykpack::bodyflags::TRACE_TAIL;
+            }
         }
 
-        Ok(())
+        // Since there's a one-to-one mapping between MIR and SIR blocks, we know how many SIR
+        // blocks we will need and can allocate empty SIR blocks ahead of time.
+        let blocks = vec![
+            ykpack::BasicBlock {
+                stmts: Default::default(),
+                term: ykpack::Terminator::Unreachable,
+            };
+            num_blocks
+        ];
+
+        Self { func: ykpack::Body { symbol_name, blocks, flags } }
     }
 
-    /// Given an llvm::BasicBlock returns the equivalent ykpack::BasicBlock in the SIR.
-    fn get_sir_block(&mut self, bb: *const BasicBlock) -> &mut ykpack::BasicBlock {
-        let (sirfuncidx, sirblockidx) = self.llvm_blocks[&bb];
-        let sir_func = &mut self.funcs[sirfuncidx];
-        &mut sir_func.blocks[sirblockidx.index()]
+    /// Returns true if there are no basic blocks.
+    pub fn is_empty(&self) -> bool {
+        self.func.blocks.len() == 0
     }
 
-    /// Set the current position of builder to `pos`. Equivalent to LLVMPositionBuilderBefore.
-    pub fn position_before(&mut self, builder: *const Builder, bb: *const BasicBlock, pos: usize) {
-        self.builders.insert(builder, (bb, pos));
+    /// Appends a statement to the specified basic block.
+    fn append_stmt(&mut self, bb: ykpack::BasicBlockIndex, stmt: ykpack::Statement) {
+        self.func.blocks[usize::try_from(bb).unwrap()].stmts.push(stmt);
     }
 
-    /// Set the current position of builder to the end of `bb`. Equivalent to
-    /// LLVMPositionBuilderAtEnd.
-    pub fn position_at_end(&mut self, builder: *const Builder, bb: *const BasicBlock) {
-        let sir_block = self.get_sir_block(bb);
-        let pos = sir_block.stmts.len();
-        self.builders.insert(builder, (bb, pos));
+    /// Sets the terminator of the specified block.
+    pub fn codegen_terminator(
+        &mut self,
+        bb: ykpack::BasicBlockIndex,
+        mir_term: &mir::Terminator<'_>,
+    ) {
+        let term = &mut self.func.blocks[usize::try_from(bb).unwrap()].term;
+        // We should only ever replace the default unreachable terminator assigned at allocation time.
+        debug_assert!(*term == ykpack::Terminator::Unreachable);
+
+        // FIXME: Nothing is implemented yet.
+        *term = ykpack::Terminator::Unimplemented(format!("{:?}", mir_term.kind));
```

Comment:
This looks like a candidate for `file!` and `line!` to differentiate it from the case in the function below?

Reply:
There's no need here as one pushes a statement, and the other sets a terminator. It's easy enough to see at a glance without the need for line nubers.

I have another branch lined up where I do use that trick for terminators. As soon as I started serialising for real I realised that you can't serialise all terminators at one call site, like I have done here.

Reply:
OK.